mailpoet/assets/js/src/segments/dynamic/subscribers_counter.tsx

```diff
@@ -24,34 +24,62 @@ const SubscribersCounter: React.FunctionComponent = () => {
   const { updateSubscriberCount } = useDispatch('mailpoet-dynamic-segments-form');
 
   const serializedSegment = JSON.stringify(segment);
+  const latestRequestIdRef = useRef(1);
+  const deferredRequestRef = useRef(null);
+  const isRequestInFlight = useRef(false);
+
+  function load(loadItem: Segment): void {
+    // Don't allow multiple in-flight requests to avoid hammering the database
+    // when we'll only ever use the results of the last request
+    if (isRequestInFlight.current) {
+      deferredRequestRef.current = loadItem;
+      return;
+    }
+
+    deferredRequestRef.current = null;
+    latestRequestIdRef.current += 1;
+    const requestId = latestRequestIdRef.current;
+    isRequestInFlight.current = true;
+
+    loadCount(loadItem).then((response) => {
+      isRequestInFlight.current = false;
+      if (deferredRequestRef.current) {
+        load(deferredRequestRef.current);
+        return;
+      }
+      if (requestId !== latestRequestIdRef.current) {
+        // Don't do anything with the response because a newer request has been initiated
+        return;
+      }
+      const finished = {} as SubscriberCount;
+      finished.loading = false;
+      if (response) {
+        finished.count = response.count;
+        finished.errors = response.errors;
+      }
+      updateSubscriberCount(finished);
+    }, (errorResponse) => {
+      isRequestInFlight.current = false;
+      const finished = {} as SubscriberCount;
+      const errors = errorResponse.errors.map((error) => error.message);
+      finished.loading = false;
+      finished.count = undefined;
+      finished.errors = errors;
+      updateSubscriberCount(finished);
+    });
+  }
+
+  const debouncedLoadRef = useRef(debounce(load, 2000, { trailing: true }));
```

Comment:
My understanding is that `useCallback` expects an inline function, not a higher order function like a debounce function. There's a good discussion about it here: https://github.com/facebook/react/issues/19240#issuecomment-652945246. I was getting `error  React Hook useCallback received a function whose dependencies are unknown. Pass an inline function instead  react-hooks/exhaustive-deps`.

In my branch that only includes the debounce logic, I had to use `useCallback` for `load` as well as pass in updateSubscriberCount as a paramater (otherwise it would need to be in the dependency array and it was causing `load` to be different every time defeating the purpose), and then `useMemo` for `debouncedLoad`, in order to avoid an eslint error: https://github.com/mailpoet/mailpoet/blob/ca0d7b16227bd4ee022bf9787d2d7015a3c56e33/mailpoet/assets/js/src/segments/dynamic/subscribers_counter.tsx#L34-L34

Here's one other place where we had to ignore a linting rule for a debounced function: https://github.com/mailpoet/mailpoet/blob/ca0d7b16227bd4ee022bf9787d2d7015a3c56e33/mailpoet/assets/js/src/form_editor/blocks/html/edit.jsx#L49-L49

Maybe it would be best to do that here as well so I can use the better `useCallback` approach? It just seems weird to have to ignore a rule like that.

Reply:
Oh, right, I missed that. If we don't expect `load` to be ever redefined even on a new component mount, then we can keep the `useRef`, I think.